### PR TITLE
make ERC20 contract upgradeable; add required dependencies; fix a typo

### PR DESCRIPTION
**Pull request by GoodluckH**
Note that some stylish changes were made automatically by my linter.

### CODE REVIEW

File: contracts/ArrowToken.sol [hunk 1]
```diff
@@ -1,29 +1,42 @@
...
-    /// @notice This function assigns msg.sender as the token admin and mints them initial supply
-    constructor(uint256 initialSupply, string memory _name, string memory _symbol)
-        ERC20(_name, _symbol)
-    {
-        _mint(msg.sender, initialSupply);
-    }
+    /** 
+        @notice Mint an `_initialSupply` amount of tokens upon contract creation
+    */
+    function initialize(uint256 _initialSupply) public initializer {
```
Comment (BrassLion) [inline contracts/ArrowToken.sol]:
```suggestion
    function initialize(uint256 _initialSupply, string memory _name, string memory _symbol) public initializer {
```
We shouldn't remove the functionality of the old constructor in the new init function.

File: contracts/ArrowToken.sol [hunk 2]
```diff
@@ -1,29 +1,42 @@
...
-    function mint(address to, uint256 amount) public onlyOwner
-    {
-        _mint(to, amount);
+    /**     
+        @notice Mint and issue new tokens to a specified address
+        @param _to the address to issue new tokens to
+        @param _amount the amount of tokens to be mint and transfered to `_to` 
+   */
+    function mint(address _to, uint256 _amount) public onlyOwner {
+        _mint(_to, _amount);
```
Comment (BrassLion) [inline contracts/ArrowToken.sol:41]: Let's do the minimum viable change. There's no need to alter the mint function in this PR. We can add documentation in a separate PR.

File: contracts/ArrowVestingBase.sol [hunk 3]
```diff
@@ -1,22 +1,23 @@
...
         uint64 durationSeconds
     ) public initializer {
-        __VestingWallet_init(beneficiaryAddress, startTimestamp, durationSeconds);
-    } 
+        __VestingWallet_init(
+            beneficiaryAddress,
+            startTimestamp,
+            durationSeconds
+        );
+    }
```
Comment (BrassLion) [inline contracts/ArrowVestingBase.sol:18]: Style changes should really be made in a separate PR.

File: contracts/ArrowVestingFactory.sol [hunk 4]
```diff
@@ -34,7 +33,16 @@ contract ArrowVestingFactory {
...
+            beneficiaryAddress,
+            startTimestamp,
+            durationSeconds
+        );
+        emit NewVestingAgreement(
+            beneficiaryAddress,
+            startTimestamp,
+            durationSeconds,
+            payable(clone)
+        );
```
Comment (BrassLion) [inline contracts/ArrowVestingFactory.sol]: As above. Style changes in a separate PR.

File: contracts/test/ArrowToken2.sol [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    UUPSUpgradeable
+{
+    // new variable for testing purpose
+    uint256 public myVal;
+
+    // implement the UUPS interface
+    function _authorizeUpgrade(address) internal override onlyOwner {}
+
+    /** 
+        @notice Introduce a new varabile (for testing purpose)
```
Comment (BrassLion) [inline contracts/test/ArrowToken2.sol]:
```suggestion
        @notice Introduce a new variable to test upgradability.
```

**Review by BrassLion — CHANGES_REQUESTED** [review]
The upgrade tests should be moved over to this PR and I'll give it another pass after that.

**Comment by GoodluckH** [comment]
It seems like I would have to include the entire test suite here because all of them needs to call "arrowToken", which, in this PR, is upgradeable. 

lmk if this is good to go.

**Review by BrassLion — CHANGES_REQUESTED** [review]
Let's rebase these changes on top of `main`, fix the conflicts, and take the review from there.

File: contracts/ArrowToken.sol [hunk 6]
```diff
@@ -1,29 +1,37 @@
...
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
 
 /**
-    @title Arrow Token implementation
- */
-contract ArrowToken is ERC20, Ownable {
+    @title Arrow Token implementation that is upgradeable
+    @notice Create ARROW tokens of an `initialSupply` amount
```
Comment (BrassLion) [inline contracts/ArrowToken.sol]:
```suggestion
    @title Arrow Token Implementation
```

File: contracts/ArrowToken.sol [hunk 7]
```diff
@@ -1,29 +1,37 @@
...
+    @dev If we were to upgrade the initializer by introducing new variables, we need to create a normal function like `initializeV2` (https://forum.openzeppelin.com/t/uups-proxies-tutorial-solidity-javascript/7786/53), and make sure to include all existing variables from the current version of the contract to avoid storage collision
```
Comment (BrassLion) [inline contracts/ArrowToken.sol]:
```suggestion
```
Move this comment to the initializer function, since it's more relevant in that scope.

File: contracts/ArrowToken.sol [hunk 8]
```diff
@@ -1,29 +1,37 @@
...
+{
+    // implement the UUPS interface
+    function _authorizeUpgrade(address) internal override onlyOwner {}
 
-    /// @notice This function assigns msg.sender as the token admin and mints them initial supply
-    constructor(uint256 initialSupply, string memory _name, string memory _symbol)
-        ERC20(_name, _symbol)
-    {
-        _mint(msg.sender, initialSupply);
-    }
```
Comment (BrassLion) [inline contracts/ArrowToken.sol:18]: We can't outright delete the constructor, as this may leave the contract uninitialized and open to attacks:

https://docs.openzeppelin.com/upgrades-plugins/1.x/writing-upgradeable#initializing_the_implementation_contract

File: contracts/test/ArrowToken2.sol [hunk 9]
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: MIT
```
Comment (BrassLion) [inline contracts/test/ArrowToken2.sol]: Let's call this file `ArrowTokenV2` to match the class contained within it.

File: package.json [hunk 10]
```diff
@@ -34,11 +34,13 @@
     "@nomiclabs/hardhat-web3": "^2.0.0",
     "@openzeppelin/contracts": "^4.5.0",
     "@openzeppelin/contracts-upgradeable": "^4.5.1",
+    "chai": "^4.3.6",
```
Comment (BrassLion) [inline package.json]: I don't think chai needs to be added to dependencies right? It's fine in `devDependencies`.

File: test/upgrade.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    arrowToken = await upgrades.deployProxy(
+      Token,
+      [ethers.BigNumber.from(1_000_000), name, symbol],
+      { kind: "uups" }
+    );
+    await arrowToken.deployed();
+    TokenV2 = await ethers.getContractFactory("ArrowTokenV2");
+    arrowTokenV2 = await upgrades.upgradeProxy(arrowToken, TokenV2, {
+      call: { fn: "initializeV2", args: [52] },
+    });
```
Comment (BrassLion) [inline test/upgrade.js:40]: It's weird to me that we're doing the upgrade in the `beforeEach` call, when the upgrade is the thing we're specifically testing here.

Could we move this into the test flow and make sure it succeeds with an `expect`?

File: test/upgrade.js [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+    it("V2 should have a different address from V1 and the proxy contract", async function () {
+      expect(arrowTokenV2.address).not.equal(
+        await upgrades.erc1967.getImplementationAddress(arrowToken.address)
+      );
+      expect(arrowTokenV2.address).not.equal(
+        await upgrades.erc1967.getImplementationAddress(arrowTokenV2.address)
+      );
+    });
```
Comment (BrassLion) [inline test/upgrade.js:105]: Can we add a test to make sure that all token holder's balances are successfully preserved after the upgrade?

Comment (BrassLion) [reply test/upgrade.js:105]: And that they can interact and send tokens successfully through the upgraded contract?

**Review by BrassLion — CHANGES_REQUESTED** [review]

File: contracts/ArrowToken.sol (hunk 6 above)
Comment (BrassLion) [reply contracts/ArrowToken.sol]: These docs are more useful when localised on the functions that control them, like `initialize`.

File: contracts/ArrowToken.sol (hunk 8 above)
Comment (GoodluckH) [reply contracts/ArrowToken.sol:18]: I think we just need to include the below codes rather than the explicit constructor.
```
/// @custom:oz-upgrades-unsafe-allow constructor
constructor() initializer {}
```
https://forum.openzeppelin.com/t/security-advisory-initialize-uups-implementation-contracts/15301

File: test/upgrade.js (hunk 11 above)
Comment (GoodluckH) [reply test/upgrade.js:40]: What are we expecting then? As long as it doesn't throw and that the following passes, we would know that it's been upgraded successfully:
```js    
it("Should have the correct value for state variable `myVal`", async function () {
      assert.strictEqual(await arrowTokenV2.myVal(), 52, "myVal value");
    });
```

File: contracts/ArrowToken.sol (hunk 8 above)
Comment (BrassLion) [reply contracts/ArrowToken.sol:18]: Yeah, that's what I was referring to. We need a constructor that initialises the contract.

File: test/upgrade.js (hunk 11 above)
Comment (BrassLion) [reply test/upgrade.js:40]: To me it looks like we don't need the `beforeEach` because each of the previous `it` statements could mimic the upgrade process instead, which would be easier to reason about.

File: test/upgrade.js [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    );
+    await arrowToken.deployed();
+    
+    // Transfer all of owner's tokens to signer1
+    let senderBalance = ethers.BigNumber.from(await arrowToken.balanceOf(owner.address));
+    amountTransferredToSigner1 = senderBalance;
+    await arrowToken.transfer(signer1.address, amountTransferredToSigner1);
+    expect(await arrowToken.balanceOf(owner.address)).to.equal(
+      senderBalance - amountTransferredToSigner1
+    );
```
Comment (BrassLion) [inline test/upgrade.js:35]: Why are we doing this in the `beforeEach`? It seems unnecessary for the vast number of tests.

File: contracts/test/ArrowTokenV2.sol [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+
+import "@openzeppelin/contracts-upgradeable/token/ERC20/ERC20Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+
+/**
+    @title Arrow Token implementation that is upgradeable
+    @notice Create ARROW tokens of an `initialSupply` amount
+    @dev If we were to upgrade the initializer by introducing new variables, we need to create a normal function like `initializeV2` (https://forum.openzeppelin.com/t/uups-proxies-tutorial-solidity-javascript/7786/53), and make sure to include all existing variables from the current version of the contract to avoid storage collision
```
Comment (BrassLion) [inline contracts/test/ArrowTokenV2.sol]: Might be more useful to let the reader know why this exists, ie as a mock to test that contracts upgrade.

File: contracts/ArrowToken.sol [hunk 15]
```diff
@@ -1,29 +1,39 @@
...
-        Mints new tokens.
+        _mint(msg.sender, _initialSupply);
+    }
 
-        Only accepts calls from the contract owner.
-     */
-    function mint(address to, uint256 amount) public onlyOwner
-    {
-        _mint(to, amount);
+    function mint(address _to, uint256 _amount) public onlyOwner {
```
Comment (BrassLion) [inline contracts/ArrowToken.sol:40]: Can you add the documentation back to this function? There's no need to remove it.

**Review by BrassLion — APPROVED** [review]
LGTM
